File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.vb
Imports System

Namespace Tests.Diagnostics
    Public Interface IFace
        Sub Method6(a As Integer()()) ' Noncompliant {{Make this method private or simplify its parameters to not use multidimensional/jagged arrays.}}
'           ^^^^^^^
    End Interface

    Public MustInherit Class Base
        MustOverride Sub Method5(a As Integer()()) ' Noncompliant
    End Class

    Public Class PublicMethodWithMultidimensionalArray
        Inherits Base
        Implements IFace
        Public Sub Method(a As Integer)
        End Sub
        Public Sub MethodX(a As Integer())
        End Sub

        Public Sub Method1(a As Integer()()) ' Noncompliant
        End Sub
        Sub Method2(a As Integer()()) ' Noncompliant
        End Sub
        Sub Method2b(a As Integer(,)) ' Noncompliant
        End Sub

        Private Sub Method3(a As Integer()())
        End Sub

        Function Method4(a As Integer()()) As Integer ' Noncompliant
            Return 1
        End Function

        Overrides Sub Method5(a As Integer()()) ' Compliant
        End Sub

        Private Sub Method6(a As Integer()()) Implements IFace.Method6 ' Compliant, interface implementation
            Throw New NotImplementedException()
        End Sub

        Public Sub Method7(ParamArray a As Integer()()) ' Compliant
        End Sub

        Public Sub Method8(ParamArray a As Integer()()()) ' Noncompliant
        End Sub

        Sub Method9(ParamArray a As Integer()(,)) ' Compliant
        End Sub

        Sub Method10 ' Compliant
            Console.WriteLine("Hello, world!")
        End Sub

        Public Sub Metho11(a As Integer()(), ParamArray b As Integer()()) ' Noncompliant
        End Sub

        Sub Method12(a As Integer(,)()) ' Noncompliant
        End Sub

    End Class
    Friend Class Other

        Public Sub Method1(a As Integer()()) ' Compliant
        End Sub

    End Class

    Public Class Constructors
        Public Sub New(A As Integer()()) ' Noncompliant {{Make this constructor private or simplify its parameters to not use multidimensional/jagged arrays.}}
'                  ^^^
        End Sub

        Public Sub New(A As Integer(,)) ' Noncompliant
        End Sub

        Public Sub New(ParamArray A As Integer())
        End Sub

        Public Sub New(ParamArray A As Integer()()()) ' Noncompliant
        End Sub

        Public Sub New(I As Integer)
        End Sub
    End Class

End Namespace

File: analyzers/tests/SonarAnalyzer.Test/TestCases/PublicMethodWithMultidimensionalArray.cs
public interface IFace
{
    void Method5(int[,] a); //Noncompliant
//       ^^^^^^^
}

public abstract class Base
{
    public abstract void Method4(int[,] a); //Noncompliant {{Make this method private or simplify its parameters to not use multidimensional/jagged arrays.}}
}

public class PublicMethodWithMultidimensionalArray : Base, IFace
{
    public void Method(int a)
    {
    }
    public void MethodX(int[] a)
    {
    }
    public void Method1(int[][] a) //Noncompliant
    {
    }
    void Method2(int[][] a) //Compliant
    {
    }
    public void Method3(int[,] a) //Noncompliant
    {
    }

    public override void Method4(int[,] a) //Compliant, overrides
    {
    }

    public void Method5(int[,] a) //Compliant, implements interface
    {
    }

    void Method6(params int[][,] a) // Compliant
    {
    }

    public void Method7(params int[][] a) // Compliant
    {
    }

    public void Method8(params int[][][] a) // Noncompliant
    {
    }

    public void Method9(int[][] a, params int[][] b) // Noncompliant
    {
    }

    public void Method10(int[,][] a) // Noncompliant
    {
    }
}

internal class Other
{
    public void Method1(int[][] a) // Compliant, class is internal
    {
    }
}

// https://github.com/SonarSource/sonar-dotnet/issues/8083
namespace Repro_8083
{
    public class Constructors
    {
        public Constructors(int[][] a) { }          // Noncompliant {{Make this constructor private or simplify its parameters to not use multidimensional/jagged arrays.}}
        public Constructors(int[,] a) { }           // Noncompliant
        public Constructors(params int[] a) { }     // Compliant, params of int
        public Constructors(params int[][][] a) { } // Noncompliant
        public Constructors(int i) { }              // Compliant, not a multi-dimensional array
    }
}

public static class ExtensionMethod
{
    public static void Method1<T>(this T[,] dataMatrix, int a) { }
    public static void Method2<T>(this T data, T[,] dataMatrix) { } // Noncompliant
}
